### DISCUSSION TITLE
Derby database connection not possible

### DISCUSSION THREAD

**Discussion by TimKuchenbuch** [quarkusio/quarkus#24347] [Q&A]
Hi all, 
I'm building a quarkus application that should connect  to a embedded derby database in order to store the application data. 

The driver is not able to use the provided URL to connect to the database `jdbc:derby:memory:myDB` . For test purposes i implanted a example project. Her i use only a memory database.

# application.properties
```
quarkus.datasource.db-kind=derby
quarkus.datasource.jdbc.url=jdbc:derby:memory:myDB
```

# build.gradle (dependencies)
```
dependencies {
    implementation enforcedPlatform("${quarkusPlatformGroupId}:${quarkusPlatformArtifactId}:${quarkusPlatformVersion}")
    implementation 'io.quarkus:quarkus-arc'
    implementation 'io.quarkus:quarkus-resteasy'

    implementation 'io.quarkus:quarkus-hibernate-orm'
    implementation 'io.quarkus:quarkus-jdbc-derby'

    testImplementation 'io.quarkus:quarkus-junit5'
    testImplementation 'io.rest-assured:rest-assured'
}
```

# Error message 
```
WARN: Datasource '<default>': Driver does not support the provided URL: jdbc:derby:memory:myDB
März 16, 2022 1:25:13 PM org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator initiateService
WARN: HHH000342: Could not obtain connection to query metadata
java.sql.SQLException: Driver does not support the provided URL: jdbc:derby:memory:myDB
	at io.agroal.pool.ConnectionFactory.connectionSetup(ConnectionFactory.java:226)
	at io.agroal.pool.ConnectionFactory.createConnection(ConnectionFactory.java:210)
	at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:513)
	at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:494)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at io.agroal.pool.util.PriorityScheduledExecutor.beforeExecute(PriorityScheduledExecutor.java:75)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

**Comment by lion78at** [comment]
In my opinion, the problem is that the embedded driver is missing in the Derby library. I tried to specify it directly - i.e. quarkus.datasource.jdbc.driver=org.apache.derby.jdbc.EmbeddedDriver -, but this doesn't work, because it is not included in the library.

**Comment by mcarbonneaux** [comment]
> Driver does not support the provided URL

i use quarkus 3.4.3 and i've the same probleme with embeded file jdbc url...

> In my opinion, the problem is that the embedded driver is missing in the Derby library. I tried to specify it directly - i.e. quarkus.datasource.jdbc.driver=org.apache.derby.jdbc.EmbeddedDriver -, but this doesn't work, because it is not included in the library.

after searching, i found that quarkus pom not use the correct dependency,  they use the network version only of derby, and the driver autoconfig of quarkus autoselect the network driver version only of derby.... 

plus the derby network driver not send comprehensible error message to advertize the user of the nature of the error...
in place of `Driver does not support the provided URL` they must say the you use the network only version....and must explaine the you need to add the good dependency to correcte the problem... i think a bugzilla on the derby project need to be issued....

in v10.14 the `org.apache.derby.jdbc.EmbeddedDriver` are in `org.apache.derby:derby` ....  and in v10.15+ he has migrated in `org.apache.derby:derbyshared`.... but quarkus `io.quarkus:quarkus-jdbc-derby` use v10.14.2.0 and use only the `org.apache.derby:derbyclient` (the network only version)... and i think is only tested with the network version....

https://github.com/quarkusio/quarkus/blob/44f9b08b795722bdfd65c6ecc0369ccafb70867d/extensions/jdbc/jdbc-derby/deployment/src/main/java/io/quarkus/jdbc/derby/deployment/JDBCDerbyProcessor.java#L30

https://github.com/quarkusio/quarkus/blob/44f9b08b795722bdfd65c6ecc0369ccafb70867d/extensions/jdbc/jdbc-derby/runtime/pom.xml#L25-L28

in v10.14.2.0 is ok for me when set `quarkus.datasource.jdbc.driver` to `org.apache.derby.jdbc.EmbeddedDriver` and manualy add this dependecy to have `org.apache.derby.jdbc.EmbeddedDriver` (in maven syntax)
```
    <!-- https://mvnrepository.com/artifact/org.apache.derby/derby -->
    <dependency>
      <groupId>org.apache.derby</groupId>
      <artifactId>derby</artifactId>
      <version>10.14.2.0</version>
    </dependency>
    <!-- https://mvnrepository.com/artifact/org.apache.derby/derbyclient -->
    <dependency>
      <groupId>org.apache.derby</groupId>
      <artifactId>derbyclient</artifactId>
      <version>10.14.2.0</version>
    </dependency>

```


_**PS: I think that quarkus must add the `org.apache.derby:derbytool` and `org.apache.derby:derby` in `org.apache.derby:derby` to permit to choose between network and embed version and modify documentation to explain that need to set quarkus.datasource.jdbc.driver=org.apache.derby.jdbc.EmbeddedDriver to use embed version. or use this by default and explain the reverse to configure the network version (i think most of the user of derby with quarkus use it as embed than network).**_


in v10.15.2.0 i need to add this dependcy to compile:
```
    <!-- https://mvnrepository.com/artifact/org.apache.derby/derbytools -->
    <dependency>
      <groupId>org.apache.derby</groupId>
      <artifactId>derbytools</artifactId>
      <version>10.15.2.0</version>
    </dependency>
    <dependency>
      <groupId>org.apache.derby</groupId>
      <artifactId>derbyshared</artifactId>
      <version>10.15.2.0</version>
    </dependency>
```

but i got : java.lang.NullPointerException: Cannot invoke "java.sql.Driver.acceptsURL(String)" because the return value of "org.apache.derby.jdbc.EmbeddedDriver.getDriverModule()

in v10.16.1.1 i need the same new dependency as v10.15 but i got:
class not found exception on org.apache.derby.shared.common.error.StandardException

**_i think v10.15+ are buggy... and need issue to be emited on bugzilla of derby project_**
